src/main/java/me/jellysquid/mods/hydrogen/mixin/chunk/MixinWorldChunk.java

```diff
@@ -3,41 +3,54 @@
-import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.gen.Accessor;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
-import java.util.function.Consumer;
 
 @Mixin(WorldChunk.class)
-public class MixinWorldChunk {
-    @Shadow
-    @Final
-    private ChunkSection[] sections;
+public abstract class MixinWorldChunk extends Chunk {
+
+    // This constructor is meaningless, just to make the compiler happy
+    public MixinWorldChunk(ChunkPos pos, UpgradeData upgradeData, HeightLimitView heightLimitView, Registry<Biome> biome, long inhabitedTime, @Nullable ChunkSection[] sectionArrayInitializer, @Nullable BlendingData blendingData) {
+        super(pos, upgradeData, heightLimitView, biome, inhabitedTime, sectionArrayInitializer, blendingData);
+    }
 
     @Inject(method = "<init>(Lnet/minecraft/world/World;Lnet/minecraft/util/math/ChunkPos;" +
-            "Lnet/minecraft/world/biome/source/BiomeArray;Lnet/minecraft/world/chunk/UpgradeData;" +
-            "Lnet/minecraft/world/TickScheduler;Lnet/minecraft/world/TickScheduler;J[" +
-            "Lnet/minecraft/world/chunk/ChunkSection;Ljava/util/function/Consumer;)V",at = @At("RETURN"))
-    private void reinit(World world, ChunkPos pos, BiomeArray biomes, UpgradeData upgradeData,
-                        TickScheduler<Block> blockTickScheduler, TickScheduler<Fluid> fluidTickScheduler,
-                        long inhabitedTime, ChunkSection[] sections, Consumer<WorldChunk> loadToWorldConsumer,
+            "Lnet/minecraft/world/chunk/UpgradeData;" +
+            "Lnet/minecraft/world/tick/ChunkTickScheduler;Lnet/minecraft/world/tick/ChunkTickScheduler;" +
+            "J[Lnet/minecraft/world/chunk/ChunkSection;" +
+            "Lnet/minecraft/world/chunk/WorldChunk$EntityLoader;Lnet/minecraft/world/gen/chunk/BlendingData;)V",
+            at = @At("RETURN"))
+    private void reinit(World world, ChunkPos pos,
+                        UpgradeData upgradeData,
+                        ChunkTickScheduler<Block> blockTickScheduler, ChunkTickScheduler<Fluid> fluidTickScheduler,
+                        long inhabitedTime, @Nullable ChunkSection[] sectionArrayInitializer,
+                        @Nullable WorldChunk.EntityLoader entityLoader, @Nullable BlendingData blendingData,
                         CallbackInfo ci) {
+
         // Upgrading a ProtoChunk to a WorldChunk might result in empty sections being copied over
         // These simply waste memory, and the WorldChunk will return air blocks for any absent section without issue.
-        for (int i = 0; i < this.sections.length; i++) {
-            if (ChunkSection.isEmpty(this.sections[i])) {
-                this.sections[i] = null;
+        for (int i = 0; i < sectionArray.length; i++) {
+            if (sectionArray[i].isEmpty()) {
+                sectionArray[i] = null;
```

Comment:
This optimization is possibly not applicable anymore because in WorldChunk we find multiple uses of `sectionArray.empty()` without a prior check for `null`, it's not immediately clear to me how we could mixin our way into intercepting the individual uses without rewritting the whole class.